bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/http/RateLimitedHttpClient.java

```diff
@@ -0,0 +1,130 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.http.internal.http;
+
+import java.net.URI;
+import java.util.concurrent.*;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.jetty.client.api.AuthenticationStore;
+import org.eclipse.jetty.client.api.Request;
+
+/**
+ * The {@link RateLimitedHttpClient} is a wrapper for a Jetty HTTP client that limits the number of requests by delaying
+ * the request creation
+ *
+ * @author Jan N. Klug - Initial contribution
+ */
+@NonNullByDefault
+public class RateLimitedHttpClient {
+    private static final int MAX_QUEUE_SIZE = 1000; // maximum queue size
+    private HttpClient httpClient;
+    private int delay = 0; // in ms
+    private final ScheduledExecutorService scheduler;
+    private final LinkedBlockingQueue<RequestQueueEntry> requestQueue = new LinkedBlockingQueue<>();
+
+    private @Nullable ScheduledFuture<?> processJob;
+
+    public RateLimitedHttpClient(HttpClient httpClient, ScheduledExecutorService scheduler) {
+        this.httpClient = httpClient;
+        this.scheduler = scheduler;
+    }
+
+    /**
+     * Stop processing the queue and clear it
+     */
+    public void shutdown() {
+        stopProcessJob();
+        requestQueue.forEach(queueEntry -> queueEntry.future.completeExceptionally(new InterruptedException()));
+    }
+
+    /**
+     * Set a new delay
+     * 
+     * @param delay in ms between to requests
+     */
+    public void setDelay(int delay) {
+        if (delay < 0) {
+            throw new IllegalArgumentException("Delay needs to be larger or equal to zero");
+        }
+        this.delay = delay;
+        stopProcessJob();
+        if (delay != 0) {
+            processJob = scheduler.scheduleWithFixedDelay(this::processQueue, 0, delay, TimeUnit.MILLISECONDS);
+        }
+    }
+
+    /**
+     * Set the HTTP client
+     *
+     * @param httpClient secure or insecure Jetty http client
+     */
+    public void setHttpClient(HttpClient httpClient) {
+        this.httpClient = httpClient;
+    }
+
+    /**
+     * Create a new request to the given URL respecting rate-limits
+     *
+     * @param finalUrl the request URL
+     * @return a CompletableFuture that completes with the request
+     */
+    public CompletableFuture<Request> newRequest(URI finalUrl) {
+        // if no delay is set, return a completed CompletableFuture
+        if (delay == 0) {
+            return CompletableFuture.completedFuture(httpClient.newRequest(finalUrl));
+        }
+        if (requestQueue.size() >= MAX_QUEUE_SIZE) {
+            return CompletableFuture.failedFuture(new IllegalStateException("Maximum queue size exceeded."));
+        }
+        CompletableFuture<Request> future = new CompletableFuture<>();
+        requestQueue.add(new RequestQueueEntry(finalUrl, future));
+        return future;
```

Comment:
You can let the queue do the capacity check instead which would simplify this a bit.
```suggestion
        CompletableFuture<Request> future = new CompletableFuture<>();
        if (!requestQueue.offer(new RequestQueueEntry(finalUrl, future)) {
            future.completeExceptionally(new RejectedExecutionException("Maximum queue size exceeded."));
        }
        return future;
```